In [1]:
import pandas as pd

READING THE CSV FILES AND CREATING DATAFRAMES

In [2]:
order_lines_dt = pd.read_csv("orderlines.csv")
orders_dt = pd.read_csv("orders.csv")
products_dt = pd.read_csv("products.csv")
brands_dt = pd.read_csv("brands.csv")

CREATING COPIES I CAN WORK ON

In [3]:
order_lines = order_lines_dt.copy()
orders = orders_dt.copy()
products = products_dt.copy()
brands = brands_dt.copy()

# EXPLORATION OF THE DATASETS 

## 1.(order_lines_dt) 

In [4]:
#JUST EXPLORING
order_lines

#A LOT OF 0 IN THE COLUMN PRODUCT_ID, LETS SEE IF THE COLUMN IS USEFULL IN ANYWAY


,id,id_order,product_id,product_quantity,sku,unit_price,date
0,1119109,299539,0,1,OTT0133,18.99,2017-01-01 00:07:19
1,1119110,299540,0,1,LGE0043,399.00,2017-01-01 00:19:45
2,1119111,299541,0,1,PAR0071,474.05,2017-01-01 00:20:57
3,1119112,299542,0,1,WDT0315,68.39,2017-01-01 00:51:40
4,1119113,299543,0,1,JBL0104,23.74,2017-01-01 01:06:38
...,...,...,...,...,...,...,...
293978,1650199,527398,0,1,JBL0122,42.99,2018-03-14 13:57:25
293979,1650200,527399,0,1,PAC0653,141.58,2018-03-14 13:57:34
293980,1650201,527400,0,2,APP0698,9.99,2018-03-14 13:57:41
293981,1650202,527388,0,1,BEZ0204,19.99,2018-03-14 13:58:01


In [5]:
order_lines.info()
#THIS DATASET HAS:
#NO NULL VALUES THANK GOD

#THE COLUMN UNIT_PRICE IS AN OBJECT SHOULD BE TRANSFORMED INTO FLOAT

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


In [6]:
#LOOKING FOR DUPLICATES
order_lines.duplicated().sum()
#THERE IS NO DUPLICATES IN THIS DATASET

np.int64(0)

In [7]:
#LOOKING AT THE NUMBER OF UNIQUE VALUES
order_lines.nunique()
#IN THE COLUMN ID_ORDER NOT, MEANING SOME ORDERS HAVE MULTIPALE PRODUCTS

#IN THE COLUMN PRODUCT_ID, ONLY 1 IS UNIQUE MEANING PROBABLY THE COLUMN IS NOT USEFLL

#IN THE COLUMN PRODUCT_QUAN:
    #SOME CUSTOMERS BUY VERY DIFFERENT QUANTATIES
    #THERE MAY BE BULK ORDERS
    #POSSIBLY WHOLESALE OR BUSSINES CUSTOMERS
    #THERE MAYBE OUTLIERS
    #DATA ERROR
    
#IN THE COLUMN SKU THERE IS 7951 UNIQUE VALUES MEANING HIGH PRODUCT VARIETY


id                  293983
id_order            204855
product_id               1
product_quantity        67
sku                   7951
unit_price           11329
date                251631
dtype: int64

In [8]:
order_lines["product_id"].unique()
#THE COLUMN PRODUCT ID IS MOSTLY 0

array([0])

In [9]:
sorted(order_lines["product_quantity"].unique())
#THERE ARE SOME BIG OULIERS IT COULD BE:
    #BUSINESS BY WHOLESALE
    #cORPORATE CUSTOMERS
    #DATA ERROR

[np.int64(1),
 np.int64(2),
 np.int64(3),
 np.int64(4),
 np.int64(5),
 np.int64(6),
 np.int64(7),
 np.int64(8),
 np.int64(9),
 np.int64(10),
 np.int64(11),
 np.int64(12),
 np.int64(13),
 np.int64(14),
 np.int64(15),
 np.int64(16),
 np.int64(17),
 np.int64(18),
 np.int64(19),
 np.int64(20),
 np.int64(21),
 np.int64(22),
 np.int64(23),
 np.int64(24),
 np.int64(25),
 np.int64(26),
 np.int64(27),
 np.int64(28),
 np.int64(29),
 np.int64(30),
 np.int64(31),
 np.int64(32),
 np.int64(33),
 np.int64(34),
 np.int64(35),
 np.int64(36),
 np.int64(37),
 np.int64(39),
 np.int64(41),
 np.int64(46),
 np.int64(47),
 np.int64(50),
 np.int64(51),
 np.int64(54),
 np.int64(55),
 np.int64(57),
 np.int64(60),
 np.int64(66),
 np.int64(69),
 np.int64(70),
 np.int64(72),
 np.int64(80),
 np.int64(89),
 np.int64(92),
 np.int64(97),
 np.int64(98),
 np.int64(99),
 np.int64(100),
 np.int64(125),
 np.int64(126),
 np.int64(164),
 np.int64(192),
 np.int64(200),
 np.int64(201),
 np.int64(555),
 np.int64(800),
 np.int64(

In [10]:
order_lines["product_quantity"].value_counts().sort_index()
#THE PRODUCT QUANTITY (999) APPEARS 2 TIMES INTERESTING
    #ITS LIKE SOMEONE HAVE BEEN WAITING FOR A SALE TO BUY 

product_quantity
1      277058
2       12711
3        1728
4        1251
5         343
        ...  
200         1
201         1
555         1
800         1
999         2
Name: count, Length: 67, dtype: int64

In [11]:
#LOOKING AT HIGH QUANTITY VALUES
order_lines.loc[order_lines_dt["product_quantity"] > 100]


,id,id_order,product_id,product_quantity,sku,unit_price,date
27779,1180010,323959,0,126,ADN0039,34.99,2017-02-14 10:21:12
40813,1204788,335057,0,201,THU0029,80.99,2017-03-14 15:25:53
53860,1228150,346221,0,999,APP1190,55.99,2017-04-14 21:50:52
57306,1234111,349133,0,555,APP0665,70.99,2017-04-24 10:20:13
57796,1234924,349475,0,800,KIN0137,7.49,2017-04-25 09:59:00
68712,1254032,358747,0,999,SEV0028,19.99,2017-05-24 14:51:58
136675,1388261,417536,0,200,TRK0009,29.99,2017-10-25 15:02:39
204637,1500715,464858,0,192,APP1662,519.00,2017-12-17 15:53:04
246048,1574262,496172,0,164,EVU0013,19.99,2018-01-22 16:14:42
285492,1637611,522075,0,125,XDO0047,25.99,2018-03-06 10:07:54


In [12]:
#MEGING ORDERS AND ORDER_LINES TO LOOK FURTHER INTO THE QUANTITY VS TOTAL PAID 
orders_orderlines = pd.merge(orders,
                             order_lines,
                             left_on="order_id",
                             right_on="id_order",
                             how="inner")



In [13]:
order_lines

,id,id_order,product_id,product_quantity,sku,unit_price,date
0,1119109,299539,0,1,OTT0133,18.99,2017-01-01 00:07:19
1,1119110,299540,0,1,LGE0043,399.00,2017-01-01 00:19:45
2,1119111,299541,0,1,PAR0071,474.05,2017-01-01 00:20:57
3,1119112,299542,0,1,WDT0315,68.39,2017-01-01 00:51:40
4,1119113,299543,0,1,JBL0104,23.74,2017-01-01 01:06:38
...,...,...,...,...,...,...,...
293978,1650199,527398,0,1,JBL0122,42.99,2018-03-14 13:57:25
293979,1650200,527399,0,1,PAC0653,141.58,2018-03-14 13:57:34
293980,1650201,527400,0,2,APP0698,9.99,2018-03-14 13:57:41
293981,1650202,527388,0,1,BEZ0204,19.99,2018-03-14 13:58:01


In [14]:
orders_orderlines.loc[orders_orderlines["product_quantity"] > 100 ]

,order_id,created_date,total_paid,state,id,id_order,product_id,product_quantity,sku,unit_price,date
28131,323959,2017-02-14 10:21:12,4408.69,Shopping Basket,1180010,323959,0,126,ADN0039,34.99,2017-02-14 10:21:12
41153,335057,2017-03-14 15:25:53,16360.00,Shopping Basket,1204788,335057,0,201,THU0029,80.99,2017-03-14 15:25:53
54192,346221,2017-04-14 21:50:52,55930.71,Shopping Basket,1228150,346221,0,999,APP1190,55.99,2017-04-14 21:50:52
57681,349133,2017-04-24 10:20:13,39399.84,Shopping Basket,1234111,349133,0,555,APP0665,70.99,2017-04-24 10:20:13
58204,349475,2017-04-25 09:59:00,5995.91,Place Order,1234924,349475,0,800,KIN0137,7.49,2017-04-25 09:59:00
69082,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254032,358747,0,999,SEV0028,19.99,2017-05-24 14:51:58
137066,417536,2017-10-25 15:02:39,5997.98,Shopping Basket,1388261,417536,0,200,TRK0009,29.99,2017-10-25 15:02:39
204969,464858,2017-12-17 15:53:04,99648.10,Shopping Basket,1500715,464858,0,192,APP1662,519.00,2017-12-17 15:53:04
246217,496172,2018-01-22 16:14:01,3278.43,Shopping Basket,1574262,496172,0,164,EVU0013,19.99,2018-01-22 16:14:42
285385,522075,2018-03-06 10:07:54,3248.70,Shopping Basket,1637611,522075,0,125,XDO0047,25.99,2018-03-06 10:07:54


In [15]:
orders_orderlines.loc[orders_orderlines["order_id"] == 358747]
#EVERYTHING SEEMS TO BEE CORRECT WITH THE QUANTITY

,order_id,created_date,total_paid,state,id,id_order,product_id,product_quantity,sku,unit_price,date
69082,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254032,358747,0,999,SEV0028,19.99,2017-05-24 14:51:58
69083,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254033,358747,0,19,APP1818,2.999.99,2017-05-24 14:53:54
69084,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254034,358747,0,9,LIBRO,0.00,2017-05-24 14:53:54
69085,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254035,358747,0,2,APP1876,3.836.99,2017-05-24 14:54:13
69086,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254036,358747,0,2,APP1914,3.494.99,2017-05-24 14:54:20
69087,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254037,358747,0,1,APP1910,3.722.99,2017-05-24 14:54:27
69088,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254038,358747,0,1,APP1900,3.722.99,2017-05-24 14:54:44
69089,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254039,358747,0,46,SEV0030,19.99,2017-05-24 14:56:46
69090,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254040,358747,0,1,TRA0026,179.99,2017-05-24 14:58:55
69091,358747,2017-05-24 14:51:58,100738.45,Shopping Basket,1254041,358747,0,1,TRA0028,559.99,2017-05-24 14:59:00


In [16]:
order_lines["date"].min()

'2017-01-01 00:07:19'

In [17]:
order_lines["date"].max()
#TIME RANGE IS 14 MONTHS

'2018-03-14 13:58:36'

## 2.(orders_dt) 

In [18]:
orders

,order_id,created_date,total_paid,state
0,241319,2017-01-02 13:35:40,44.99,Cancelled
1,241423,2017-11-06 13:10:02,136.15,Completed
2,242832,2017-12-31 17:40:03,15.76,Completed
3,243330,2017-02-16 10:59:38,84.98,Completed
4,243784,2017-11-24 13:35:19,157.86,Cancelled
...,...,...,...,...
226904,527397,2018-03-14 13:56:38,42.99,Place Order
226905,527398,2018-03-14 13:57:25,42.99,Shopping Basket
226906,527399,2018-03-14 13:57:34,141.58,Shopping Basket
226907,527400,2018-03-14 13:57:41,19.98,Shopping Basket


In [19]:
orders.info()
#THIS DATASET HAS:
# 5 NULL VALUES IN THE COLUMN TOTAL_PAID

#THE COLUMN CREATED_DATE IS AN OBJECT IT SHOULD BE TRANSFORMED INTO DATETIME


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


In [20]:
#LOOKING FOR DUPLICATES
orders.duplicated().sum()
#THERE IS NO DUPLICATES IN THIS DATASET

np.int64(0)

In [21]:
#LOOKING FOR NUMBER OF UNIQUE VALUES
orders.nunique()



order_id        226909
created_date    224828
total_paid       31236
state                5
dtype: int64

In [22]:
orders["created_date"].min()

'2017-01-01 00:07:19'

In [23]:
orders["created_date"].max()
#TIMERANGE IS THE SAME AS ORDER_LINES

'2018-03-14 13:58:36'

In [24]:
(orders["total_paid"].isna().sum()
 / 
 orders.shape[0]*100).round(5)
#THE MISSING VALUES ARE ONLY 0.0022% FROM FROM THE WHOLE DATASET
#I RECOMMEND DROPPING THEM OR FILLING IT OUT WITH "WAITNG"

np.float64(0.0022)

## 3.(products_dt) 

In [25]:
products

,sku,name,desc,price,promo_price,in_stock,type
0,RAI0007,Silver Rain Design mStand Support,Aluminum support compatible with all MacBook,59.99,499.899,1,8696
1,APP0023,Apple Mac Keyboard Keypad Spanish,USB ultrathin keyboard Apple Mac Spanish.,59,589.996,0,13855401
2,APP0025,Mighty Mouse Apple Mouse for Mac,mouse Apple USB cable.,59,569.898,0,1387
3,APP0072,Apple Dock to USB Cable iPhone and iPod white,IPhone dock and USB Cable Apple iPod.,25,229.997,0,1230
4,KIN0007,Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM,2GB RAM Mac mini and iMac (2006/07) MacBook Pr...,34.99,31.99,1,1364
...,...,...,...,...,...,...,...
19321,BEL0376,Belkin Travel Support Apple Watch Black,compact and portable stand vertically or horiz...,29.99,269.903,1,12282
19322,THU0060,"Enroute Thule 14L Backpack MacBook 13 ""Black",Backpack with capacity of 14 liter compartment...,69.95,649.903,1,1392
19323,THU0061,"Enroute Thule 14L Backpack MacBook 13 ""Blue",Backpack with capacity of 14 liter compartment...,69.95,649.903,1,1392
19324,THU0062,"Enroute Thule 14L Backpack MacBook 13 ""Red",Backpack with capacity of 14 liter compartment...,69.95,649.903,0,1392


In [26]:
products.info()
#THIS DATA SET HAS:
# 7 NULL VALUES IN THE COLUMN DESC

# 46 NULL VALUES IN THE COLUMN PRICE

#THE COLUMN PRICE AND PROMO_PRICE ARE OBJECTS SHOULD BE TRANSFORMED INTO FLOATS

#TYPE CAN REMAIN AN OBJECT AT FIRST 


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


In [27]:
#LOOKING FOR DUPLICATES 
products.duplicated().sum()
#THERE IS 8746 DUPLICATES IN THIS DATASET

np.int64(8746)

In [28]:
products = products_dt.drop_duplicates()

In [29]:
products[products.duplicated()]

,sku,name,desc,price,promo_price,in_stock,type


In [30]:
#LOOKING FOR NUMBER OF UNIQUE VALUES 
products.nunique()
#THE COLUMN IN_STOCK HAS EITHER 0 OR 1 WHICH I FIND WEIRD
#THE COLUMN TYPE HAS 126 UNIQUE VALUES MEANING WE CAN IDENTIFY THOSE AS CATEGORIES I THINK 

sku            10579
name           10373
desc            7098
price           2690
promo_price     4614
in_stock           2
type             126
dtype: int64

In [31]:
#LOOKING INTO TWO DECIMALS PROMO_PRICES 
products[products["promo_price"].str.count(r"\.") > 1]
 

,sku,name,desc,price,promo_price,in_stock,type
50,APP0367,Apple Mini DisplayPort to DVI Adapter Mac dual...,Adapter Mini Display Port to DVI dual channel ...,119,1.119.976,0,1325
51,APP0344,"Apple Thunderbolt Display 27 ""Monitor Mac",Monitor Display 27-inch Apple Thunderbolt (MC9...,1149,10.449.923,0,1296
66,MAK0008,Maclocks theft case iPad 2 3 and 4 black with ...,Case antitheft iPad 2 3 and 4 polycarbonate ro...,120,1.199.957,0,12635403
67,MAK0007,Maclocks theft case iPad 2 3 and 4 transparent...,Case antitheft iPad 2 3 and 4 polycarbonate ro...,120,1.079.961,0,12635403
97,MAK0014,Maclocks safety housing Kiosk iPad 2 3 and 4 b...,Holder and housing iPad 2 3 and 4 aluminum and...,164.99,1.649.896,0,1216
...,...,...,...,...,...,...,...
19300,REP0412,Rear Camera Repair iPhone 7 Plus,It is including parts and labor for iPhone 7 Plus,119.99,1.199.897,0,"1,44E+11"
19317,REP0403,iPad LCD screen repair,Repair service including parts and labor for iPad,159.99,1.599.898,0,"1,44E+11"
19318,REP0402,iPad touch screen repair,Repair service including parts and labor for iPad,139.99,1.399.897,0,"1,44E+11"
19319,KNO0032,"Knomo MacBook Pro Beauchamp Backpack 14 ""Black",Backpack thin nylon mesh internal compartment ...,179,1.699.905,1,1392


In [32]:
#LOOKING INTO BASE_PRICES WITH TWO DECIMALS 
products[products["price"].str.count(r"\.") > 1]


,sku,name,desc,price,promo_price,in_stock,type
665,CRU0015-2,Crucial memory Mac 16GB (2x8GB) SO-DIMM DDR3 1...,RAM 16GB (2x8GB) 135V MacBook Pro iMac (2012/2...,1.639.792,1.629.894,1,1364
792,APP0672,Apple iPhone 5S 16GB Space Gray,New iPhone 5S 16G Libre (ME432Y / AB).,4.694.994,4.694.994,0,NaN
797,APP0673,Apple iPhone 5S 16GB Silver,New Free iPhone 5S 16GB (ME433Y / A).,4.090.042,4.090.042,0,NaN
827,PAC0339,NewerTech miniStack 4TB Hard Drive Mac,External Box Hard Drive Mac + 4TB.,2.199.791,2.199.901,0,11935397
885,PAC0376,OWC Mercury Elite Pro Dual Thunderbolt + 8TB,RAID outer box 35 inch SATA connection Thunder...,5.609.698,5.549.895,0,11935397
...,...,...,...,...,...,...,...
19312,REP0424,Input repair Headphones iPad,Repair service including parts and labor for iPad,6.999.003,69.99,0,"1,44E+11"
19313,REP0421,iPad charging connector repair,Repair service including parts and labor for iPad,6.999.003,69.99,0,"1,44E+11"
19314,REP0416,iPad front camera repair,Repair service including parts and labor for iPad,6.999.003,69.99,0,"1,44E+11"
19315,REP0413,repair rear camera iPad,Repair service including parts and labor for iPad,6.999.003,69.99,0,"1,44E+11"


In [33]:
#LOOKING IF THERE IS PROMO PRICES THAT IS BIGGER THAN BASE_PRICE
products.loc[products["promo_price"] > products["price"]]
#IT LOOKS LIKE THE PROMO PRICE IS MULTIPLICATED BY 10

,sku,name,desc,price,promo_price,in_stock,type
15,MOS0021,Clearguard Moshi MacBook Pro and Air,Keyboard Protector MacBook Pro 13-inch Retina ...,24.95,249.889,0,13835403
17,HGD0007,"Henge Docks Docking Station MacBook Pro 13 """,Dock and support with multiple connections for...,69.99,699.864,0,12995397
19,APP0234,Apple Dock Connector to VGA,Dock Connector to VGA IOS.,35,350.005,0,13955395
23,MOB0007,Mobee Magic Charger Charger Magic Mouse,Wireless Charger for Magic Mouse,49.99,499.851,0,12355400
39,JMO0026,Just Mobile Lazy Couch Support Mac and iPad,Mac and iPad small lift stand.,19.95,199.892,0,8696
...,...,...,...,...,...,...,...
19312,REP0424,Input repair Headphones iPad,Repair service including parts and labor for iPad,6.999.003,69.99,0,"1,44E+11"
19313,REP0421,iPad charging connector repair,Repair service including parts and labor for iPad,6.999.003,69.99,0,"1,44E+11"
19314,REP0416,iPad front camera repair,Repair service including parts and labor for iPad,6.999.003,69.99,0,"1,44E+11"
19315,REP0413,repair rear camera iPad,Repair service including parts and labor for iPad,6.999.003,69.99,0,"1,44E+11"


In [34]:
order_lines

,id,id_order,product_id,product_quantity,sku,unit_price,date
0,1119109,299539,0,1,OTT0133,18.99,2017-01-01 00:07:19
1,1119110,299540,0,1,LGE0043,399.00,2017-01-01 00:19:45
2,1119111,299541,0,1,PAR0071,474.05,2017-01-01 00:20:57
3,1119112,299542,0,1,WDT0315,68.39,2017-01-01 00:51:40
4,1119113,299543,0,1,JBL0104,23.74,2017-01-01 01:06:38
...,...,...,...,...,...,...,...
293978,1650199,527398,0,1,JBL0122,42.99,2018-03-14 13:57:25
293979,1650200,527399,0,1,PAC0653,141.58,2018-03-14 13:57:34
293980,1650201,527400,0,2,APP0698,9.99,2018-03-14 13:57:41
293981,1650202,527388,0,1,BEZ0204,19.99,2018-03-14 13:58:01


In [35]:
#MERGING THE ORDERS_ORDERLINES AND THE PRODUCTS TI DIG FURTHER INTO PRICES WITH TWO DECIMALS
orders_products = pd.merge(orders_orderlines,
                           products,
                           on="sku",
                           how="left")
orders_products.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293840 entries, 0 to 293839
Data columns (total 17 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   order_id          293840 non-null  int64  
 1   created_date      293840 non-null  object 
 2   total_paid        293834 non-null  float64
 3   state             293840 non-null  object 
 4   id                293840 non-null  int64  
 5   id_order          293840 non-null  int64  
 6   product_id        293840 non-null  int64  
 7   product_quantity  293840 non-null  int64  
 8   sku               293840 non-null  object 
 9   unit_price        293840 non-null  object 
 10  date              293840 non-null  object 
 11  name              292425 non-null  object 
 12  desc              292412 non-null  object 
 13  price             292211 non-null  object 
 14  promo_price       292425 non-null  object 
 15  in_stock          292425 non-null  float64
 16  type              29

In [36]:
#LOOKING FURTHER INTO PRICES WITH TWO DECIMALS
orders_products.loc[orders_products["price"]
                    .astype(str).str.count(r"\.") > 1,
                    ["order_id", "unit_price", "price", "name"]]
#NOT IN ALL CASES I CAN FIND A PATTERN TO SOLVE SO:
    #INSTEAD OF CREATING PRICES I RECOMMEND TO DROP THEM

,order_id,unit_price,price,name
24,251688,152.95,1.990.002,Apple TV 32GB 4K
86,265954,14.53,1.728.001,Kingston DataTraveler SE9 32Gb USB 3.0 G2
126,278926,17.56,2.499.013,iOttie iTap magnetic car holder for iPhone
603,299860,457.89,5.608.689,QNAP TS-451A NAS l 8GB RAM
648,299902,23.74,2.499.013,iOttie iTap magnetic car holder for iPhone
...,...,...,...,...
293287,526844,219.00,21.900.032,Apple TV 4K 64GB
293529,527099,299.99,2.999.905,Station 3 Caldigit Thunderbolt Thunderbolt out...
293709,527286,1.159.00,115.900.092,Apple iPhone 64GB X Silver
293774,527342,1.329.00,13.290.011,Apple iPhone X 256GB Silver


In [37]:
orders_products.loc[orders_products["unit_price"]
                    .str.count(r"\.") > 1,
                    ["order_id", "unit_price", "price"]]
#THERE IS A PATTERN WE SHOULD REMOVE THE FIRST DECIMAL

,order_id,unit_price,price
53,257847,1.367.11,1489
58,258985,2.258.01,2459
65,259668,1.087.33,1089
74,262016,2.905.59,3279
101,268372,1.131.44,1205.99
...,...,...,...
293709,527286,1.159.00,115.900.092
293752,527321,3.497.00,4519
293754,527324,3.075.00,3319
293774,527342,1.329.00,13.290.011


In [38]:
#cols = ["price", "promo_price", "unit_price"]

#for col in cols:
   # orders_products[col] = (
        #orders_products[col]
        #.astype(str)
        #.str.replace(r"\.(?=.*\.)", "", regex=True)
        #.astype(float))

#AFTER TRYING TO CLEAN THE COLUMN PROMO_PRICE OR PRICE IN THE DATASETS WITHOUT LOSING VALUES I CONCLUDED:
   #THAT IM GONNA END UP WITH NOT SUCH TRUSTWORTHY VALUES THAT WOULD AFFECT MY ANALYSIS
   #THATS WHY I DECIDED ON ROMVING ANYTHING THAT IS NOT RELAIBLE

In [39]:
orders_products["price"].sort_values()
#THAT SHOWED ME BAD VALUES AFTER TRYING TO CLEAN IT WITHOUT LOSING VALUES

212699    1.090.004
219632    1.090.004
211537    1.090.004
213619    1.090.004
23407     1.169.795
            ...    
267094          NaN
270351          NaN
276871          NaN
283906          NaN
285010          NaN
Name: price, Length: 293840, dtype: object

In [40]:
orders_products["unit_price"].max()

'999.99'

In [41]:
products.loc[products['desc'].isna(), :]
#I CAN ADD THE NAME TO THE MISSING VALUES

,sku,name,desc,price,promo_price,in_stock,type
16126,WDT0211-A,"Open - Purple 2TB WD 35 ""PC Security Mac hard ...",NaN,107,814.659,0,1298
16128,APP1622-A,Open - Apple Smart Keyboard Pro Keyboard Folio...,NaN,1.568.206,1.568.206,0,1298
17843,PAC2334,Synology DS718 + NAS Server | 10GB RAM,NaN,566.35,5.659.896,0,12175397
18152,KAN0034-A,Open - Kanex USB-C Gigabit Ethernet Adapter Ma...,NaN,29.99,237.925,0,1298
18490,HTE0025,Hyper Pearl 1600mAh battery Mini USB Mirror an...,NaN,24.99,22.99,1,1515
18612,OTT0200,OtterBox External Battery Power Pack 20000 mAHr,NaN,79.99,56.99,1,1515
18690,HOW0001-A,Open - Honeywell thermostat Lyric zonificador ...,NaN,199.99,1.441.174,0,11905404


In [42]:
products.loc[(products.price.str.contains(r"\d+\.\d+\.\d+"))
                    |
                    (products.price.str.contains(r"\d+\.\d{3,}")), :].shape[0]
#THERE ARE 596 ROWS WITH TWO DECIMALS OR 3 NUMBERS AFTER THE SECOND DECIMAL 

542

In [43]:
price_problems_number = products.loc[(products.price.str.contains(r"\d+\.\d+\.\d+"))
                                        |
                                        (products.price.str.contains(r"\d+\.\d{3,}")), :].shape[0]
price_problems_number#

542

In [44]:
 round((price_problems_number / products.shape[0]) * 100,2)
 #THE PERCANTAGE OF HARD TO SAVE VALUES IN PRICE IS 3,08%


5.12

In [45]:
promo_problems_number = products.loc[(products.promo_price.str.contains(r"\d+\.\d+\.\d+"))
                                        |
                                        (products.promo_price.str.contains(r"\d+\.\d{3,}")), :].shape[0]
promo_problems_number
#THERE ARE 17547 ROWS WITH TWO DECIMALS OR 3 NUMBERS AFTER THE SECOND DECIMAL

9734

In [46]:
round(((promo_problems_number / products.shape[0]) * 100), 2)
#THE PERCANTAGE OF HARD TO SAVE VALUES IN PROMO_PRICE IS 93.68%

92.0

## 4.(brand_dt) 

In [47]:
#JUST EXPLORING
brands_dt

,short,long
0,8MO,8Mobility
1,ACM,Acme
2,ADN,Adonit
3,AII,Aiino
4,AKI,Akitio
...,...,...
182,XOO,Xoopar
183,XRI,X-Rite
184,XTO,Xtorm
185,ZAG,ZaggKeys


In [48]:
brands_dt.info()
#EVERYTHNIG SEEMS TO BE FINE WITH THIS DATASET

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187 entries, 0 to 186
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   short   187 non-null    object
 1   long    187 non-null    object
dtypes: object(2)
memory usage: 3.1+ KB


In [49]:
#LOOKING FOR DUPLICATES
brands_dt.duplicated().sum()
#THERE IS NO DUPLICATES IN THIS DATASET

np.int64(0)

In [50]:
#LOOKING FOR THE NUMBERS OF UNIQUE VALUES
brands_dt.nunique()
#SHORT AND LONG DOESNT MATCH WHICH IS WEIRD

short    187
long     181
dtype: int64

In [51]:
brands_dt[brands_dt["long"].duplicated(keep=False)]
#THERE ARE UNKNOWN VALUES IN THE COLUMN LONG

,short,long
6,AP2,Apple
7,APP,Apple
17,BOS,Bose
19,CAD,Bose
37,ENV,Unknown
67,JYB,Jaybird
70,KEN,Jaybird
80,LIB,Unknown
94,MOP,Mophie
98,MUJ,Mophie


# CLEANING AND MANUPULATING THE DATASETS

## 1.(order_lines_dt)

In [52]:
#TURNING THE COLUMN UNIT_PRICE TO FLOAT
order_lines["unit_price"] = pd.to_numeric(order_lines["unit_price"], errors="coerce")

order_lines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                293983 non-null  int64  
 1   id_order          293983 non-null  int64  
 2   product_id        293983 non-null  int64  
 3   product_quantity  293983 non-null  int64  
 4   sku               293983 non-null  object 
 5   unit_price        257814 non-null  float64
 6   date              293983 non-null  object 
dtypes: float64(1), int64(4), object(2)
memory usage: 15.7+ MB


In [53]:
#CHANGING THE COLUMN DATE DATATYPE
order_lines["date"] = pd.to_datetime(order_lines["date"])

order_lines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        257814 non-null  float64       
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 15.7+ MB


In [54]:
#DROPING THE COLUMN PRODUCT_ID BECAUSE ITS NOT USEFULL
order_lines = order_lines.drop(columns=["product_id"])


In [55]:
#CHECKING IF IT WORKED
order_lines

,id,id_order,product_quantity,sku,unit_price,date
0,1119109,299539,1,OTT0133,18.99,2017-01-01 00:07:19
1,1119110,299540,1,LGE0043,399.00,2017-01-01 00:19:45
2,1119111,299541,1,PAR0071,474.05,2017-01-01 00:20:57
3,1119112,299542,1,WDT0315,68.39,2017-01-01 00:51:40
4,1119113,299543,1,JBL0104,23.74,2017-01-01 01:06:38
...,...,...,...,...,...,...
293978,1650199,527398,1,JBL0122,42.99,2018-03-14 13:57:25
293979,1650200,527399,1,PAC0653,141.58,2018-03-14 13:57:34
293980,1650201,527400,2,APP0698,9.99,2018-03-14 13:57:41
293981,1650202,527388,1,BEZ0204,19.99,2018-03-14 13:58:01


## 2.(orders)

In [56]:
#CHANGING THE COLUMN CREATED DATE TO DATETYPE
orders["created_date"]= pd.to_datetime(orders["created_date"])
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226909 non-null  int64         
 1   created_date  226909 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226909 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 6.9+ MB


## 3.(products_dt)

In [57]:
#CLEANING THE COLUMN PRICE FROM NUMBERS WITH 2 DECIMALS OR WITH 3 NUMBERS AFTER THE LAST DECIMAL
products = products.loc[~((products.price.str.contains(r"\d+\.\d+\.\d+"))|(products.price.str.contains(r"\d+\.\d{3,}"))), :]
products.sample(10)

,sku,name,desc,price,promo_price,in_stock,type
12650,OWC0166,OWC SSD expansion kit Aura 6G 960GB iMac 21.5 ...,Upgrade kit 960GB SSD iMac 21.5-inch / 27-inch...,568.99,5.035.851,0,12215397
492,OWC0022,SSD expansion kit OWC Aura Pro 6G 480GB MacBoo...,480GB SSD expansion for Mid 2012 MacBook Air 1...,379.99,3.005.846,1,12215397
2547,OWC0059-A,(Open) OWC Aura Pro 6G 480GB SSD for MacBook P...,480GB SSD hard drive for MacBook Pro Retina 20...,339.99,3.055.945,0,1298
2538,FCM0025-8,Mac FCM memory 32GB (8x4GB) 1333MHz DDR3 DIMM,RAM 32GB (8x4GB) Mac Pro (July 2010).,247.92,2.309.902,0,1364
524,SEV0023,HDD or SSD installation service Mac Pro,HDD or SSD installation on your Mac Pro + Data...,49.99,499.851,0,20642062
10625,APP1461,Apple HDMI to DVI Adapter,IPhone iPad iPod adapter and AV cable.,35,330.003,0,1230
16243,PAC2118,"Apple iMac 27 ""Core i5 Retina 5K 38GHz | 16GB ...",IMac desktop computer 27 inch Retina 5K RAM 16...,3079,28.190.045,0,"5,74E+15"
18906,APP2377,"Apple Macbook Pro 13 ""Core i7 25GHz | 16GB | 5...",New MacBook Pro 13-inch Core i7 25GHz with 16G...,2599,24.430.045,0,"1,02E+12"
17358,WAC0164-A,Open - Wacom Intuos Creative Comic Pen & Touch...,Graphics Tablet touch with pencil included for...,99.99,709.538,0,1298
18750,SNS0019-A,Open - Sonos Playbar Black Speaker,Reconditioned Soundbar HiFi,799,6.171.127,0,5398


In [58]:
#CHANGNING THE COLUMN PRICE TO FLOAT
products["price"] = pd.to_numeric(products["price"])
products.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10038 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          10038 non-null  object 
 1   name         10038 non-null  object 
 2   desc         10032 non-null  object 
 3   price        9992 non-null   float64
 4   promo_price  10038 non-null  object 
 5   in_stock     10038 non-null  int64  
 6   type         9992 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 627.4+ KB


In [59]:
#DROPING THE COLUMN PROMO_PRICE BECAUSE:
    #92% IS NOT TRUST WORTHY
    #I HAVE THE UNIT_PRICE THAT IS TRUST WORTHY INSTEAD
products = products.drop(columns=["promo_price"])
products 

,sku,name,desc,price,in_stock,type
0,RAI0007,Silver Rain Design mStand Support,Aluminum support compatible with all MacBook,59.99,1,8696
1,APP0023,Apple Mac Keyboard Keypad Spanish,USB ultrathin keyboard Apple Mac Spanish.,59.00,0,13855401
2,APP0025,Mighty Mouse Apple Mouse for Mac,mouse Apple USB cable.,59.00,0,1387
3,APP0072,Apple Dock to USB Cable iPhone and iPod white,IPhone dock and USB Cable Apple iPod.,25.00,0,1230
4,KIN0007,Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM,2GB RAM Mac mini and iMac (2006/07) MacBook Pr...,34.99,1,1364
...,...,...,...,...,...,...
19321,BEL0376,Belkin Travel Support Apple Watch Black,compact and portable stand vertically or horiz...,29.99,1,12282
19322,THU0060,"Enroute Thule 14L Backpack MacBook 13 ""Black",Backpack with capacity of 14 liter compartment...,69.95,1,1392
19323,THU0061,"Enroute Thule 14L Backpack MacBook 13 ""Blue",Backpack with capacity of 14 liter compartment...,69.95,1,1392
19324,THU0062,"Enroute Thule 14L Backpack MacBook 13 ""Red",Backpack with capacity of 14 liter compartment...,69.95,0,1392


In [60]:
products["desc"] = (products["desc"].fillna(products["name"]))
products.loc[products["desc"].isna(),:]

,sku,name,desc,price,in_stock,type


In [61]:
#DROPPING THE NAN VALUES 
products = products.dropna()
products.isna().sum()

sku         0
name        0
desc        0
price       0
in_stock    0
type        0
dtype: int64

In [62]:
#DROPING DUPLICATES
products = products.drop_duplicates()
products.duplicated().sum()

np.int64(0)

In [63]:
products = products.drop_duplicates(subset="sku")

4.(orders_products)

In [64]:
#MERGING ORDERS WITH ORDER_LINES 
orders_orderlines_dt = pd.merge(orders,
                             order_lines,
                             left_on="order_id",
                             right_on="id_order",
                             how="inner")

In [65]:
#DROPING THE MISSING VALUES IN THE MERGED DATASET SO ALL ORDER ID CAN BE UNIQUE 
orders_orderlines_dt = orders_orderlines_dt.dropna()
orders_orderlines_dt.info()

<class 'pandas.core.frame.DataFrame'>
Index: 257606 entries, 0 to 293748
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   order_id          257606 non-null  int64         
 1   created_date      257606 non-null  datetime64[ns]
 2   total_paid        257606 non-null  float64       
 3   state             257606 non-null  object        
 4   id                257606 non-null  int64         
 5   id_order          257606 non-null  int64         
 6   product_quantity  257606 non-null  int64         
 7   sku               257606 non-null  object        
 8   unit_price        257606 non-null  float64       
 9   date              257606 non-null  datetime64[ns]
dtypes: datetime64[ns](2), float64(2), int64(4), object(2)
memory usage: 21.6+ MB


In [66]:
#MERGING THE MERGED DATA SET WITH PRODUCTS
orders_products_dt = pd.merge(orders_orderlines_dt,
                           products,
                           on="sku",
                           how="left")
orders_products_dt.isna().sum()

order_id               0
created_date           0
total_paid             0
state                  0
id                     0
id_order               0
product_quantity       0
sku                    0
unit_price             0
date                   0
name                7182
desc                7182
price               7182
in_stock            7182
type                7182
dtype: int64

In [67]:
orders_products_dt.duplicated().sum()

np.int64(0)

In [68]:
orders_products_dt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 257606 entries, 0 to 257605
Data columns (total 15 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   order_id          257606 non-null  int64         
 1   created_date      257606 non-null  datetime64[ns]
 2   total_paid        257606 non-null  float64       
 3   state             257606 non-null  object        
 4   id                257606 non-null  int64         
 5   id_order          257606 non-null  int64         
 6   product_quantity  257606 non-null  int64         
 7   sku               257606 non-null  object        
 8   unit_price        257606 non-null  float64       
 9   date              257606 non-null  datetime64[ns]
 10  name              250424 non-null  object        
 11  desc              250424 non-null  object        
 12  price             250424 non-null  float64       
 13  in_stock          250424 non-null  float64       
 14  type

In [69]:
orders_products_dt

,order_id,created_date,total_paid,state,id,id_order,product_quantity,sku,unit_price,date,name,desc,price,in_stock,type
0,241319,2017-01-02 13:35:40,44.99,Cancelled,1121139,241319,1,JBL0123,44.99,2017-01-02 12:26:59,JBL T450 Bluetooth BT Headset White,Wireless headphones with folding design with 1...,49.95,1.0,5384
1,241423,2017-11-06 13:10:02,136.15,Completed,1398738,241423,1,LAC0212,129.16,2017-11-06 12:47:20,LaCie Porsche Design Desktop Drive 4TB USB 3.0...,External Hard Drive 4TB 35-inch USB 3.0 for Ma...,139.99,1.0,11935397
2,242832,2017-12-31 17:40:03,15.76,Completed,1529178,242832,1,PAR0074,10.77,2017-12-31 17:26:40,Parrot 550mAh battery for MiniDrones,550mAh rechargeable battery for Parrot minidrones,17.99,0.0,11905404
3,243330,2017-02-16 10:59:38,84.98,Completed,1181923,243330,1,OWC0074,77.99,2017-02-15 17:07:44,Mac OWC Memory 8GB 1066MHZ DDR3 SO-DIMM,8GB RAM Mac mini iMac MacBook and MacBook Pro ...,99.99,1.0,1364
4,243784,2017-11-24 13:35:19,157.86,Cancelled,1437153,243784,3,PHI0080,51.29,2017-11-24 13:27:41,Philips HUE White Ambience Pack 2 GU10 bulbs,Pack of two bulbs with different shades of whi...,59.95,1.0,11905404
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
257601,527397,2018-03-14 13:56:38,42.99,Place Order,1650198,527397,1,JBL0122,42.99,2018-03-14 13:56:38,JBL T450 BT Bluetooth Headset Black,Wireless headphones with folding design with 1...,49.95,1.0,5384
257602,527398,2018-03-14 13:57:25,42.99,Shopping Basket,1650199,527398,1,JBL0122,42.99,2018-03-14 13:57:25,JBL T450 BT Bluetooth Headset Black,Wireless headphones with folding design with 1...,49.95,1.0,5384
257603,527399,2018-03-14 13:57:34,141.58,Shopping Basket,1650200,527399,1,PAC0653,141.58,2018-03-14 13:57:34,Samsung SSD 850 expansion kit EVO 250GB + Data...,SSD upgrade kit 2008-2010 250 GB MacBook and M...,215.98,1.0,1433
257604,527400,2018-03-14 13:57:41,19.98,Shopping Basket,1650201,527400,2,APP0698,9.99,2018-03-14 13:57:41,Apple Lightning Cable Connector to USB 1m Whit...,Apple Lightning USB Cable 1 meter to charge an...,25.00,1.0,1230


In [70]:
#EXCLUDING ORDERS THAT ARE NOT COMPLETED
#PLACE ORDER ROWS ARE UNCONFIRMED TRANSACTIONS — EXCLUDED FROM ALL ANALYSIS
orders_products_dt = orders_products_dt.loc[orders_products_dt["state"] == "Completed"]
orders_products_dt["state"].value_counts()

state
Completed    58542
Name: count, dtype: int64

In [71]:
#DROPPING REMAINING NAN VALUES (2.86% OF DATASET)
orders_products_dt = orders_products_dt.dropna()
orders_products_dt.isna().sum()

order_id            0
created_date        0
total_paid          0
state               0
id                  0
id_order            0
product_quantity    0
sku                 0
unit_price          0
date                0
name                0
desc                0
price               0
in_stock            0
type                0
dtype: int64

In [72]:
#FINAL SHAPE BEFORE OUTLIER ANALYSIS
orders_products_dt.info()

<class 'pandas.core.frame.DataFrame'>
Index: 57034 entries, 1 to 257318
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   order_id          57034 non-null  int64         
 1   created_date      57034 non-null  datetime64[ns]
 2   total_paid        57034 non-null  float64       
 3   state             57034 non-null  object        
 4   id                57034 non-null  int64         
 5   id_order          57034 non-null  int64         
 6   product_quantity  57034 non-null  int64         
 7   sku               57034 non-null  object        
 8   unit_price        57034 non-null  float64       
 9   date              57034 non-null  datetime64[ns]
 10  name              57034 non-null  object        
 11  desc              57034 non-null  object        
 12  price             57034 non-null  float64       
 13  in_stock          57034 non-null  float64       
 14  type              57034 no

## OUTLIER ANALYSIS

In [73]:
#CALCULATING DISCOUNT PERCENTAGE FROM PRICE AND UNIT_PRICE
orders_products_dt["discount_%"] = (
    (orders_products_dt["price"] - orders_products_dt["unit_price"])
    / orders_products_dt["price"]
) * 100
orders_products_dt["discount_%"].describe().round(2)

count    57034.00
mean        21.48
std         18.09
min       -212.27
25%          9.11
50%         17.90
75%         28.80
max         99.98
Name: discount_%, dtype: float64

### DISCOUNT OUTLIERS

In [74]:
#FLAGGING DISCOUNT OUTLIERS
def flag_discount(row):
    if row["unit_price"] < 1:
        return "near-free (unit_price < €1)"
    elif row["discount_%"] > 75:
        return "suspicious (>75% discount)"
    elif row["discount_%"] < 0:
        return "impossible (unit_price > price)"
    else:
        return "normal"

orders_products_dt["discount_flag"] = orders_products_dt.apply(flag_discount, axis=1)
orders_products_dt["discount_flag"].value_counts()

discount_flag
normal                             54551
impossible (unit_price > price)     1629
suspicious (>75% discount)           852
near-free (unit_price < €1)            2
Name: count, dtype: int64

In [75]:
#SHOWING THE SUSPICIOUS AND NEAR-FREE ROWS
orders_products_dt.loc[
    orders_products_dt["discount_flag"] != "normal",
    ["order_id", "sku", "name", "unit_price", "price", "discount_%", "discount_flag"]
].sort_values("discount_%")

,order_id,sku,name,unit_price,price,discount_%,discount_flag
50941,347233,IFX0074,iFixit piece Earpiece Speaker iPhone 6,24.95,7.99,-212.265332,impossible (unit_price > price)
85548,379925,IFX0074,iFixit piece Earpiece Speaker iPhone 6,24.95,7.99,-212.265332,impossible (unit_price > price)
94551,388586,IFX0074,iFixit piece Earpiece Speaker iPhone 6,24.95,7.99,-212.265332,impossible (unit_price > price)
88585,382838,PRY0008,Prynt White iPhone Case Portable Printer,135.99,74.99,-81.344179,impossible (unit_price > price)
26446,324159,KIN0149-2,Mac memory Kingston 16GB (2x8GB) SO-DIMM DDR3å...,170.99,105.98,-61.341763,impossible (unit_price > price)
...,...,...,...,...,...,...,...
240239,512862,MOP0109,Mophie Powerstation Mini Universal 4000mAh Bat...,2.00,69.95,97.140815,suspicious (>75% discount)
239823,512451,MOP0107,Mophie Powerstation 4000mAh Battery Plus Mini ...,1.60,69.95,97.712652,suspicious (>75% discount)
239861,512498,MOP0107,Mophie Powerstation 4000mAh Battery Plus Mini ...,1.60,69.95,97.712652,suspicious (>75% discount)
63060,358440,SEV0018,IMac RAM installation service,0.01,19.99,99.949975,near-free (unit_price < €1)


### ZERO TOTAL_PAID ORDERS

In [76]:
#FINDING ORDERS WHERE TOTAL_PAID IS ZERO — NO LEGITIMATE SALE
zero_paid = orders_products_dt.loc[orders_products_dt["total_paid"] == 0]
print(f"Orders with total_paid = €0: {zero_paid['order_id'].nunique()}")
zero_paid[["order_id", "sku", "name", "unit_price", "total_paid", "state"]]

Orders with total_paid = €0: 2


,order_id,sku,name,unit_price,total_paid,state
195,296010,TUC0252,"Tucano Nido Hard-Shell Case MacBook Air 13 ""Black",24.99,0.0,Completed
2199,301495,TUC0277,Svolta Tucano MacBook Pro Sleeve bag / Retina ...,24.99,0.0,Completed


### DUPLICATE ORDER LINES

In [77]:
#FINDING ROWS WHERE SAME SKU APPEARS TWICE IN SAME ORDER WITH SAME PRICE
dup_mask = orders_products_dt.duplicated(
    subset=["order_id", "sku", "unit_price"], keep=False
)
dup_rows = orders_products_dt.loc[dup_mask]
print(f"Duplicate order-SKU rows: {len(dup_rows)}")
dup_rows[["order_id", "sku", "name", "unit_price", "product_quantity"]].sort_values("order_id")

Duplicate order-SKU rows: 2


,order_id,sku,name,unit_price,product_quantity
104918,398536,ALL0010,PowerCube Original USB Regleta Allocacoc wall ...,15.99,1
104919,398536,ALL0010,PowerCube Original USB Regleta Allocacoc wall ...,15.99,1


### BULK / HIGH QUANTITY ORDERS

In [78]:
#DISTRIBUTION OF PRODUCT QUANTITY
orders_products_dt["product_quantity"].describe().round(2)

count    57034.00
mean         1.13
std          0.80
min          1.00
25%          1.00
50%          1.00
75%          1.00
max         72.00
Name: product_quantity, dtype: float64

In [79]:
#ROWS WITH QUANTITY ABOVE 10 — LIKELY B2B OR BULK BUYERS
bulk = orders_products_dt.loc[orders_products_dt["product_quantity"] > 10]
print(f"Bulk rows (qty > 10): {len(bulk)} ({round(len(bulk)/len(orders_products_dt)*100, 2)}% of dataset)")
bulk[["order_id", "sku", "name", "product_quantity", "unit_price", "total_paid"]].sort_values(
    "product_quantity", ascending=False
)

Bulk rows (qty > 10): 36 (0.06% of dataset)


,order_id,sku,name,product_quantity,unit_price,total_paid
146381,438871,SAN0186,SanDisk Ultra 16GB microSDHC Memory Card A1,72,8.65,622.82
217569,495987,STA0016,Startech Mini DisplayPort to DisplayPort Cable...,50,9.99,506.48
234055,507768,SAN0110,SanDisk Ultra Flair Flash Drive 16GB USB 3.0,37,9.00,339.99
1196,300475,OTT0132,OtterBox Symmetry Alpha Glass Case + Screen Pr...,35,29.99,1049.65
139390,432242,CRU0050,Crucial MX300 275GB SSD Disk,35,77.90,2726.66
186387,468784,SEA0049,"Seagate Barracuda 1TB Hard Drive 35 ""SATA for ...",33,44.28,1471.35
6446,305651,JAW0048,Jawbone UP2 Black Rope Activity Monitor,32,29.99,959.68
142027,434662,SAM0070,512GB Samsung SSD 850 PRO,32,214.90,6876.94
144281,436894,APP0698,Apple Lightning Cable Connector to USB 1m Whit...,30,8.99,274.71
143159,435746,WAC0191,Education - Wacom Intuos Comic M Black,27,141.67,3825.13


### TOTAL_PAID vs LINE TOTAL MISMATCH

In [80]:
#COMPARING TOTAL_PAID TO UNIT_PRICE x PRODUCT_QUANTITY
#NOTE: orders_price = unit_price x product_quantity (USE orders_price FOR REVENUE)
orders_products_dt["line_total"] = (
    orders_products_dt["unit_price"] * orders_products_dt["product_quantity"]
)
orders_products_dt["paid_vs_line_diff"] = (
    orders_products_dt["total_paid"] - orders_products_dt["line_total"]
)
print(f"Rows where total_paid > line total: {(orders_products_dt['paid_vs_line_diff'] > 0).sum()} (likely shipping)")
print(f"Rows where total_paid < line total: {(orders_products_dt['paid_vs_line_diff'] < 0).sum()} (likely vouchers)")
print(f"Median difference: €{orders_products_dt['paid_vs_line_diff'].median():.2f}")
orders_products_dt[["order_id", "total_paid", "line_total", "paid_vs_line_diff"]].sort_values(
    "paid_vs_line_diff", ascending=False
).head(10)

Rows where total_paid > line total: 48333 (likely shipping)
Rows where total_paid < line total: 306 (likely vouchers)
Median difference: €6.99


,order_id,total_paid,line_total,paid_vs_line_diff
43081,339735,9987.76,63.20,9924.56
43079,339735,9987.76,429.00,9558.76
43080,339735,9987.76,429.60,9558.16
36586,333662,9423.90,6.99,9416.91
36589,333662,9423.90,23.99,9399.91
36590,333662,9423.90,26.99,9396.91
36585,333662,9423.90,32.99,9390.91
36591,333662,9423.90,46.99,9376.91
36592,333662,9423.90,99.99,9323.91
36587,333662,9423.90,246.99,9176.91


## ADDITIONAL CLEANING

In [81]:
#DROPPING ORDERS WHERE TOTAL_PAID IS ZERO — NOT A REAL TRANSACTION
before = len(orders_products_dt)
orders_products_dt = orders_products_dt.loc[orders_products_dt["total_paid"] > 0]
print(f"Dropped: {before - len(orders_products_dt)} rows | Remaining: {len(orders_products_dt)}")

Dropped: 2 rows | Remaining: 57032


In [82]:
#DROPPING ROWS WHERE UNIT_PRICE < €1 — INTERNAL SERVICE ENTRIES, NOT REAL SALES
before = len(orders_products_dt)
orders_products_dt = orders_products_dt.loc[orders_products_dt["unit_price"] >= 1]
print(f"Dropped: {before - len(orders_products_dt)} rows | Remaining: {len(orders_products_dt)}")

Dropped: 2 rows | Remaining: 57030


In [83]:
#DROPPING ROWS WITH DISCOUNT > 75% — ECONOMICALLY SUSPICIOUS CLEARANCE ITEMS
#(CONCENTRATED IN POST-HOLIDAY PERIODS, MOSTLY DISCONTINUED PRODUCTS)
before = len(orders_products_dt)
orders_products_dt = orders_products_dt.loc[orders_products_dt["discount_%"] <= 75]
print(f"Dropped: {before - len(orders_products_dt)} rows | Remaining: {len(orders_products_dt)}")

Dropped: 852 rows | Remaining: 56178


In [84]:
#DROPPING DUPLICATE ORDER LINES — SAME SKU ORDERED TWICE IN SAME ORDER AT SAME PRICE
before = len(orders_products_dt)
orders_products_dt = orders_products_dt.drop_duplicates(
    subset=["order_id", "sku", "unit_price"]
)
print(f"Dropped: {before - len(orders_products_dt)} rows | Remaining: {len(orders_products_dt)}")

Dropped: 1 rows | Remaining: 56177


In [86]:
orders_products_dt.to_csv("orders_products_dt")

In [85]:
#FINAL DATASET VERIFICATION
print(f"Final shape: {orders_products_dt.shape}")
print(f"State distribution:\n{orders_products_dt['state'].value_counts()}")
print(f"Discount range: {orders_products_dt['discount_%'].min():.1f}% to {orders_products_dt['discount_%'].max():.1f}%")
print(f"Unit price range: €{orders_products_dt['unit_price'].min():.2f} to €{orders_products_dt['unit_price'].max():.2f}")
print(f"Nulls remaining: {orders_products_dt.isna().sum().sum()}")

Final shape: (56177, 19)
State distribution:
state
Completed    56177
Name: count, dtype: int64
Discount range: -212.3% to 74.8%
Unit price range: €1.99 to €999.99
Nulls remaining: 0


## Data Quality Assessment

A data quality check was performed to review the completeness, validity, and consistency of the dataset before starting the analysis.


## Data Quality Issues

### 1. Pricing inconsistencies (important)
Around 4–5% of the rows contain cases where the unit price is higher than the base price, which should not normally happen.

**Possible reasons:**
- The base price might be outdated  
- Some products (e.g. premium brands) may follow different pricing rules  

**Action taken:**
- Excluded these rows from discount-related analysis  
- Kept them in mind for separate investigation  

---

### 2. Missing values (moderate)
Missing values were mainly found in:
- `price` (important)  
- `desc` and `type` (less important)  

**Impact:**
- Missing prices directly affect discount calculations  

**Action taken:**
- Removed rows where the price was missing or invalid  

---

### 3. Data type issues (resolved)
Some columns were stored incorrectly:
- Prices were stored as text  
- Dates were stored as objects  

**Action taken:**
- Converted `price` and `unit_price` to numeric  
- Converted `created_date` and `date` to datetime  

---

### 4. Total price inconsistencies (moderate)
Some rows show `total_paid` is lower than `unit_price × product_quantity`.

**Possible reasons:**
- Order-level discounts  
- Shipping costs added at order level  
- Vouchers reducing the total  

**Action taken:**
- Kept these rows for analysis  
- Noted as a structural limitation — use `orders_price` for revenue, not `unit_price` alone  

---

### 5. Quantity outliers — B2B bulk orders (moderate)
Most orders have small quantities (1–3 items), but a few rows show very high values (e.g. 32×, 206×, 800×).

**Finding:**
Deep analysis confirmed these are legitimate B2B transactions (schools, businesses, corporate buyers).  
Discount behavior is virtually identical to retail customers (20.9% vs 19.4% avg).  
Combined they represent only 0.35% of total revenue.

**Action taken:**
- Kept in the dataset  
- Flagged as B2B segment for separate interpretation  

---

### 6. Unconfirmed orders — Place Order state (important)
31,965 rows had the state `Place Order`, meaning the transaction was initiated but never confirmed.

**Finding:**
`Place Order` rows show significantly lower average discounts (15.5% vs 21.6%) and higher average unit prices (€474 vs €239) compared to `Completed` orders.  
Including them distorts all metrics.

**Action taken:**
- Dropped all `Place Order` rows  
- Kept only `Completed` transactions  

---

### 7. Zero revenue orders (important)
2 orders had `total_paid = €0`, meaning no revenue was recorded despite products being listed.

**Possible reasons:**
- Test orders  
- Data entry errors  
- Free promotional orders  

**Action taken:**
- Dropped these 2 rows — not legitimate sales  

---

### 8. Near-free service entries (important)
2 rows had `unit_price = €0.01` — RAM installation and SSD installation services sold at near-zero price.

**Finding:**
These are internal service entries, not real product sales. They inflate discount percentages and distort unit price analysis.

**Action taken:**
- Dropped all rows where `unit_price < €1`  

---

### 9. Extreme discount outliers (important)
931 rows had a discount percentage above 75%, reaching up to 99.98%.

**Finding:**
These are concentrated in post-holiday clearance periods (Jan 2017, Oct 2017, Feb 2018) and consist mainly of discontinued brands (Jawbone, Zagg, Griffin).  
They are economically suspicious for a standard tech retailer and skew discount analysis significantly.

**Action taken:**
- Dropped rows where `discount_% > 75%`  

---

### 10. Duplicate order lines (minor)
4 rows were exact duplicates — the same SKU appeared twice within the same order at the same price.

**Possible reasons:**
- Data entry errors  
- System duplication during order processing  

**Action taken:**
- Dropped duplicate rows on `order_id + sku + unit_price`  

---

## Final Data Preparation

To ensure reliable analysis, the following steps were applied in order:

- Filtered to `Completed` orders only — removed 31,965 unconfirmed rows  
- Removed rows with missing or invalid prices  
- Dropped `total_paid = €0` orders — 2 rows  
- Dropped `unit_price < €1` service entries — 2 rows  
- Dropped `discount_% > 75%` clearance outliers — 931 rows  
- Dropped duplicate order-SKU rows — 4 rows  
- Standardized all data types  

---

## Summary

The dataset required significant cleaning, particularly around order state filtering and discount outliers.  
After all steps, the data is consistent and reliable for pricing, discount, and revenue analysis.  
Bulk B2B orders were retained but should be treated as a separate segment during interpretation.
